File: Algorithms/Medium/RotateList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None:
            return None

        length = 1
        tail = head
        while tail.next:
            length += 1
            tail = tail.next

        tail.next = head
        k = k % length
        curr = head
        for _ in range(length - k - 1):
            curr = curr.next

        new_head = curr.next
        curr.next = None

        return new_head

File: Algorithms/Medium/RotateList.go
package medium

import "github.com/parviz-yu/LeetCode/Algorithms/helpers"

// Time complexity: O(n)
// Space complexity: O(1)
func rotateRight(head *helpers.ListNode, k int) *helpers.ListNode {
	if head == nil {
		return head
	}

	len, curr := 1, head
	for curr.Next != nil {
		curr = curr.Next
		len++
	}

	curr.Next = head
	k = k % len
	for i := 0; i < len-k-1; i++ {
		head = head.Next
	}

	curr = head.Next
	head.Next = nil

	return curr
}
